### PR TITLE
Add distribution shift robustness

### LINKED ISSUE #2: Implement data-shift robustness
inspo: https://github.com/google-deepmind/distribution_shift_framework/tree/master

### PR DESCRIPTION
**Pull request by zuzg**
Evaluating model trained on various data shifts (class proportions).

### CODE REVIEW

File: merci/evaluate.py [hunk 1]
```diff
@@ -114,3 +115,46 @@ def evaluate(self) -> float:
...
+    ) -> None:
+        """
+        Initialize the transductive model evaluator.
+
+        :param model: Classification model to evaluate
+        :param train_dataset: A tuple containing input and target data for training
+        :param test_dataset: A tuple containing input and target data for testing
+        """
+        super().__init__(model, train_dataset, test_dataset)
+        self.thresholds = [0.75, 0.5, 0.25, 0.1]
```
Comment (Tremirre) [inline merci/evaluate.py]: perhaps we could allow for optionally passing different thresholds, and setting this as a default?

File: merci/evaluate.py [hunk 2]
```diff
@@ -114,3 +115,46 @@ def evaluate(self) -> float:
...
+        """
+        x_train, y_train = self.train_dataset
+        x_test, y_test = self.test_dataset
+        self.model.fit(x_train, y_train)
+        scores = []
+        class0 = np.where(y_train == 0)[0]
+        class1 = np.where(y_train == 1)[0]
+        for cls, other_cls in zip([class0, class1], [class1, class0]):
+            for thr in self.thresholds:
+                reduced = np.random.choice(cls, np.floor(len(cls)*thr).astype(int))
```
Comment (Tremirre) [inline merci/evaluate.py]: If we use a random method here, imo for reproducibility we should add a `seed` parameter to the evaluator

File: pipeline/experiment.py [hunk 3]
```diff
@@ -46,10 +46,10 @@ def run_pipeline(self) -> None:
...
                     (x_train, y_train),
                     (x_test, y_test),
-                    lambda a, b: (
-                        self.merge_arrays(a[0], b[0]),
-                        self.merge_arrays(a[1], b[1]),
-                    ),
+                    # lambda a, b: (
+                    #     self.merge_arrays(a[0], b[0]),
+                    #     self.merge_arrays(a[1], b[1]),
+                    # ),
```
Comment (Tremirre) [inline pipeline/experiment.py]: commented code

**Review by Tremirre — APPROVED** [review]
👌

**Review by azywot — APPROVED** [review]
lgtm 👍

File: merci/evaluate.py (hunk 1 above)
Comment (Tremirre) [reply merci/evaluate.py]: setting a mutable object as a default variable introduces a very subtle [bug where all instances of this class will point to the same object](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil)

a minor thing but it's best to be avoided ;p
recommended approach: either use a tuple or default to None:
```python
 def __init__(
        ...
        thresholds: list[int] = None,
        seed: int = 0,
    ) -> None:
        super().__init__(model, train_dataset, test_dataset)
        self.thresholds = thresholds
        if self.thresholds is None:
            self.thresholds = [0.75, 0.5, 0.25, 0.1]
        self.rng = np.random.default_rng(seed)
```

**Review by Tremirre — APPROVED** [review]
